### PR TITLE
Custom expression plugin framework for UCR

### PR DESCRIPTION
**Pull request by NoahCarnahan**
Enables custom reports to easily add expression types to UCR.

**Updated interface:**
To register a new expression, simply add a tuple to the `CUSTOM_UCR_EXPRESSIONS` setting list. The first item in the tuple is the name of the expression type, the second item is a function with a signature like `conditional_expression(spec, context)` that returns an expression object. e.g.:

``` python
# settings.py

CUSTOM_UCR_EXPRESSIONS = [
    ('abt_supervisor', 'custom.abt.reports.expressions.abt_supervisor'),
]
```

TODO?: Register expressions to specific domains, instead of globally.

**Old interface:**
To register a new expression, decorate its factory function like so:

``` python
# custom/my_org/reports/expressions.py
from corehq.apps.userreports.expressions.factory import ExpressionFactory

@ExpressionFactory.register("awesome")
def _awesome_expression(spec, context):
   ...
```

and import the file containing this definition in `corehq/apps/userreports/expressions/__init__.py` to bootstrap the plugin.

### CODE REVIEW

File: corehq/apps/userreports/expressions/__init__.py [hunk 1]
```diff
@@ -1 +1,3 @@
-
+# Bootstrap plugin expressions by importing files that contain
```
Comment (dimagimon) [inline corehq/apps/userreports/expressions/__init__.py]: F401 'custom' imported but unused

File: corehq/apps/userreports/expressions/factory.py [hunk 2]
```diff
@@ -7,13 +7,74 @@
...
+    functools.partial(_simple_expression_generator, ConstantGetterSpec)
+)
+_property_name_expression = ExpressionFactory.register('property_name')(
+    functools.partial(_simple_expression_generator, PropertyNameGetterSpec)
+)
+_property_path_expression = ExpressionFactory.register('property_path')(
+    functools.partial(_simple_expression_generator, PropertyPathGetterSpec)
+)
 
+@ExpressionFactory.register("conditional")
```
Comment (dimagimon) [inline corehq/apps/userreports/expressions/factory.py]: E302 expected 2 blank lines, found 1

File: corehq/apps/userreports/tests/test_expressions.py [hunk 3]
```diff
@@ -11,6 +11,25 @@
 from corehq.apps.userreports.specs import EvaluationContext
 
 
+class ExpressionPluginTest(SimpleTestCase):
+
+    def test_custom_expression(self):
+        """Confirm that plugin framework is being bootstrapped."""
+        expression = ExpressionFactory.from_spec({
```
Comment (dimagimon) [inline corehq/apps/userreports/tests/test_expressions.py]: F841 local variable 'expression' is assigned to but never used

File: custom/abt/reports/expressions.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+class AbtSupervisorExpressionSpec(JsonObject):
+    # TODO:
+    # This class is a stub added to demonstrate UCR expression plugin usage
+    # This class will be filled out in a PR coming soon - NC 4/20/15
+    type = TypeProperty('abt_supervisor')
+
+    def __call__(self, item, context=None):
+        return item
+
+@ExpressionFactory.register("abt_supervisor")
```
Comment (dimagimon) [inline custom/abt/reports/expressions.py]: E302 expected 2 blank lines, found 1

**Comment by snopoke** [comment]
Looks good to me. 

If you want to restrict certain expressions to domains I would say the validation should happen when they save the datasource.

File: corehq/apps/userreports/expressions/__init__.py (hunk 1 above)
Comment (dimagimon) [inline corehq/apps/userreports/expressions/__init__.py]: F401 'custom' imported but unused

**Comment by NoahCarnahan** [comment]
FYI code review buddy @biyeun

**Comment by czue** [comment]
Looks nice to me.

I think is pretty subjective, but I still prefer a settings-based approach to the decorator model. I find it easier to understand exactly what's configured if it's all defined together, rather than having to grep the code to see what gets bootstrapped. It's less elegant at definition time, but that seems an acceptable tradeoff. It also allows us to configure different plugins for different environments on the same codebase (which I think this model won't be able to do).

Not going to fight about it though if you think you can make the case for decorators.

**Comment by NoahCarnahan** [comment]
@czue 

I don't think I'm necessarily attached to the decorator pattern.
I'm not sure I know exactly what you have in mind with the settings-based approach though.
This is what I'm imaging a settings-based approach might look like, similar to what you were thinking?:

```python
# settings.py

UCR_EXPRESSION_TYPES = [
    ('constant', 'corehq.apps.userreports.expressions.factory.constant_expression'),
    ...
    ('abt_supervisor', 'custom.abt.reports.expressions.abt_supervisor'),
]
```

```python
# corehq/apps/userreports/expressions/__init__.py

for type_name, factory_function_path in UCR_EXPRESSION_TYPES:
    ExpressionFactory.register(type_name, __import__(factory_function_path))
```

**Comment by czue** [comment]
yeah, something like that. I also think it's fine to ship anything generally useful in the UCR core so no need to make the whole list of built in expressions part of settings (though the decorator pattern does make that separation case a bit nicer and the "everything is a plugin" approach likely has some design benefits).

I'm thinking that if and when we support additional levels of plugin-ability we might just have 

`UCR_PLUGINS = ['mymodule.ucr']` or something along those lines. Maybe some combination of the two could actually work out nicely with that pattern (?)

Anyway, like I said, I think the mechanics of how things register/fit together are easy to change, and the fact that something works and can be used/extended makes me not care too much about changing anything if nothing feels obviously better than anything else.

File: corehq/apps/userreports/expressions/factory.py [hunk 5]
```diff
@@ -51,6 +52,21 @@ class ExpressionFactory(object):
         'root_doc': _root_doc_expression,
         'related_doc': _related_doc_expression,
     }
+    # Additional items are added to the spec_map by use of the `register` method.
+
+    @classmethod
+    def register(cls, type_name, factory_func):
+        """
+        Registers an expression factory function for the given type_name.
+        Use this method to add additional expression types to UCR.
```
Comment (dimagimon) [inline corehq/apps/userreports/expressions/factory.py]: W291 trailing whitespace

File: corehq/apps/userreports/tests/test_expressions.py [hunk 6]
```diff
@@ -11,6 +11,21 @@
...
+        """Confirm that plugin framework is being bootstrapped."""
+        ExpressionFactory.from_spec({
+            "type": "abt_supervisor"
+        })
+
+    def test_double_registration(self):
+
+        ExpressionFactory.register("foo", lambda x: x)
+        with self.assertRaises(ValueError):
+            ExpressionFactory.register("foo", lambda x: x*2)
```
Comment (dimagimon) [inline corehq/apps/userreports/tests/test_expressions.py]: E226 missing whitespace around arithmetic operator

**Comment by NoahCarnahan** [comment]
@czue agreed that it's easy to change around. I gave the settings approach a try, and I think I like it better as it only requires a single step for registration (adding to the setting) as opposed to decorating and adding an import.

I think this is good to merge.

File: corehq/apps/userreports/tests/test_expressions.py [hunk 7]
```diff
@@ -11,6 +11,21 @@
...
+class ExpressionPluginTest(SimpleTestCase):
+
+    def test_custom_expression(self):
+        """Confirm that plugin framework is being bootstrapped."""
+        ExpressionFactory.from_spec({
+            "type": "abt_supervisor"
+        })
+
+    def test_double_registration(self):
+
```
Comment (snopoke) [inline corehq/apps/userreports/tests/test_expressions.py]: Extra blank line

**Comment by snopoke** [comment]
Nice, this is quite a bit less code as well.

**Comment by czue** [comment]
:+1: cool, thanks for responding to feedback so quickly!

**Pull request merged**